src/charm.py

```diff
@@ -39,540 +55,378 @@ def __init__(self, *args):
-
-        if successful_installation:
-            self._stored.slurm_installed = True
+        if self._slurmctld_manager.install():
 
             # Store the munge_key and jwt_rsa key in the stored state.
-            # NOTE: Use leadership settings instead of stored state when
-            # leadership settings support becomes available in the framework.
-            if self._is_leader():
-                # NOTE the backup controller should also have the jwt and munge
-                #      keys configured. We should move these information to the
-                #      peer relation.
-                self._stored.jwt_rsa = self._slurm_manager.generate_jwt_rsa()
-                self._stored.munge_key = self._slurm_manager.get_munge_key()
-                self._slurm_manager.write_jwt_rsa(self.get_jwt_rsa())
+            # NOTE: Use secrets instead of stored state when secrets are supported the framework.
+            if self.model.unit.is_leader():
+                self._stored.jwt_rsa = self._slurmctld_manager.generate_jwt_rsa()
+                self._stored.munge_key = self._slurmctld_manager.get_munge_key()
+                self._slurmctld_manager.write_jwt_rsa(self.get_jwt_rsa())
+                self._slurmctld_manager.restart_munged()
+                self.unit.set_workload_version(self._slurmctld_manager.version())
+                self.slurm_installed = True
             else:
-                # NOTE: the secondary slurmctld should get the jwt and munge
-                #       keys from the peer relation here
-                logger.debug("secondary slurmctld")
-
-            # all slurmctld should restart munged here, as it would assure
-            # munge is working
-            self._slurm_manager.restart_munged()
+                logger.debug("Secondary slurmctld not supported.")
+                event.defer()
         else:
             self.unit.status = BlockedStatus("Error installing slurmctld")
             event.defer()
 
+        self._on_write_slurm_conf(event)
+
+    def _on_config_changed(self, event: ConfigChangedEvent) -> None:
+        """Perform config-changed operations."""
+        charm_config_nhc_params = self.config.get("health-check-params")
+        if (user_supplied_nhc_params := charm_config_nhc_params) != self._stored.nhc_params:
+            logger.debug("## NHC user supplied params changed, sending to slurmd.")
+            self._stored.nhc_params = user_supplied_nhc_params
+            # Send the custom NHC parameters to all slurmd.
+            self._slurmd.set_nhc_params(user_supplied_nhc_params)
+
+        write_slurm_conf = False
+        if charm_config_default_partition := self.config.get("default-partition"):
+            if charm_config_default_partition != self._stored.default_partition:
+                logger.debug("## Default partition configuration changed.")
+                self._stored.default_partition = charm_config_default_partition
+                write_slurm_conf = True
```

Comment:
the reason we can't do this is because `config.get("my-config")` can return `None` and as such, the code breaks if we try to store `None` in the stored state...

I guess we could go with `config.get("my-config", "")` .... thoughts?

Reply:
Or we could specify a default value for the config in the `charmcraft.yaml`.... so that no matter what it is a string...